torch/functional.py

```diff
@@ -1471,7 +1471,10 @@ def _lu_impl(A, pivot=True, get_infos=False, out=None):
         LU factorization succeeded for all samples!
     """
     # If get_infos is True, then we don't need to check for errors and vice versa
-    return torch._lu_with_info(A, pivot=pivot, check_errors=(not get_infos))
+    if not A.requires_grad:
+        return torch._lu_with_info(A, pivot=pivot, check_errors=(not get_infos))
+    else:
+        return A.lu(pivot=pivot, get_infos=get_infos)
```

Comment:
CI complains with the following message from `test_jit`:
```
Tried to access nonexistent attribute or method 'lu' of type 'Tensor'.:
  File "/opt/conda/lib/python3.6/site-packages/torch/functional.py", line 1477
        return torch._lu_with_info(A, pivot=pivot, check_errors=(not get_infos))
    else:
        return A.lu(pivot=pivot, get_infos=get_infos)
               ~~~~ <--- HERE
```
Weird, I expected that `tensor.py` gets compiled before `functional.py`

Reply:
I think you can solve `mypy` issues by just having the the appropriate `isinstance()` calls that will tell it when the optional argument is None or not.

For the jit test itself I'm not sure why it doesn't pick up the implementation, do we need to add a `@torch.script` (or something like that, I don't know the syntax) decorator to the implementation of `Tensor.lu` to let it know it is properly scriptable?

Reply:
Ah, I think the issue is lu_unpack which is used in forward, so there is cross-linking going on. Now I am thinking maybe it is actually easier to write ATen lu_unpack with backward for `_lu_with_info`...